### PR TITLE
Support for custom build types on Android

### LINKED ISSUE #11095: Support custom build variants on Android
I'm trying to integrate Flutter into an existing Android app, and had to disable our "beta" custom build variant otherwise Flutter refuses to compile - even if that particular variant isn't selected.

Are there any plans to support custom build variants, or should I start working on an alternative configuration switch?

## Steps to Reproduce

Add a custom variant to an existing Android app/build.gradle:
```gradle
beta {
            initWith release
            applicationIdSuffix ".beta"
            versionNameSuffix "-beta"
}
```
## Logs
```
$ flutter run
Launching lib/main.dart on Pixel in debug mode...
FAILURE: Build failed with an exception.

* Where:
Script '/home/cachapa/flutter/packages/flutter_tools/gradle/flutter.gradle' line: 195

* What went wrong:
A problem occurred evaluating root project 'android'.
> A problem occurred configuring project ':app'.
   > Build variant must be one of "debug", "profile", or "release" but was "beta"
```

## Flutter Doctor
```
$ flutter doctor
[✓] Flutter (on Linux, locale en_GB.UTF-8, channel alpha)
    • Flutter at /home/cachapa/flutter
    • Framework revision d36e2f6191 (2 weeks ago), 2017-06-21 15:09:10 -0700
    • Engine revision b0dee695ec
    • Tools Dart version 1.24.0-dev.6.7

[✓] Android toolchain - develop for Android devices (Android SDK 25.0.3)
    • Android SDK at /home/cachapa/Android/Sdk
    • Platform android-26, build-tools 25.0.3
    • Java binary at: /opt/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[✓] Android Studio (version 2.3)
    • Android Studio at /opt/android-studio
    • Gradle version 3.2
    • Java version OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[✓] IntelliJ IDEA Community Edition (version 2017.1)
    • Flutter plugin version 15.0
    • Dart plugin version 171.4694.29

[✓] Connected devices
    • Pixel • FA69S0307169 • android-arm • Android 8.0.0 (API 26)
```

### PR DESCRIPTION
**Pull request by mravn-google**
Custom build variants are now supported by the Flutter Gradle plugin. This involves making sure that Flutter dependencies are injected into Gradle configurations and tasks beyond the default Android `debug` and `release` build types.

Fixes #11095

### CODE REVIEW

File: packages/flutter_tools/gradle/flutter.gradle [hunk 1]
```diff
@@ -168,6 +167,42 @@ class FlutterPlugin implements Plugin<Project> {
         }
     }
 
+    /**
+     * Adds suitable flutter.jar compile dependencies to the specified buildType.
+     *
+     * Note: The BuildType DSL type is not public, and is therefore just Object here.
+     */
+    private void addFlutterJarCompileDependency(Project project, Object buildType) {
```
Comment (jakobr-google) [inline packages/flutter_tools/gradle/flutter.gradle]: I believe Groovy-style is to leave the type out when you can't/don't want to specify it for arguments (and use `def` when declaring untyped variables). Section 3 of http://groovy-lang.org/style-guide.html. :)

**Review by jakobr-google — APPROVED** [review]
LGTM, +1, would review again.

**Comment by Hixie** [comment]
This patch substantially affected the build_time_millis benchmark.

**Comment by FullstackJack** [comment]
How do you run these buildTypes from the Flutter CLI?

**Pull request merged**